File: frontend/src/modules/settings/config/rag.config.ts
/**
 * RAG Settings Configuration
 *
 * Defines the structure and behavior of the RAG settings page.
 */

import type { SimpleConfigSettingsConfig } from './types';
import i18n from '../../../i18n';

export const ragConfig: SimpleConfigSettingsConfig = {
  type: 'config',
  get title() { return i18n.t('settings:rag.title'); },
  get description() { return i18n.t('settings:rag.description'); },

  apiEndpoint: {
    get: '/api/rag/config',
    update: '/api/rag/config'
  },

  fields: [
    {
      type: 'preset',
      name: 'retrieval_preset',
      get label() { return i18n.t('settings:rag.field.retrievalPreset'); },
      get helpText() { return i18n.t('settings:rag.field.retrievalPreset.help'); },
      options: [
        {
          value: 'fast',
          get label() { return i18n.t('settings:rag.opt.presetFast'); },
          get description() { return i18n.t('settings:rag.opt.presetFast.desc'); },
          effects: { top_k: 3, recall_k: 10, max_per_doc: 1, score_threshold: 0.4 }
        },
        {
          value: 'balanced',
          get label() { return i18n.t('settings:rag.opt.presetBalanced'); },
          get description() { return i18n.t('settings:rag.opt.presetBalanced.desc'); },
          effects: { top_k: 5, recall_k: 20, max_per_doc: 2, score_threshold: 0.3 }
        },
        {
          value: 'deep',
          get label() { return i18n.t('settings:rag.opt.presetDeep'); },
          get description() { return i18n.t('settings:rag.opt.presetDeep.desc'); },
          effects: { top_k: 10, recall_k: 50, max_per_doc: 3, score_threshold: 0.2 }
        }
      ]
    },
    {
      type: 'select',
      name: 'vector_store_backend',
      get label() { return i18n.t('settings:rag.field.vectorStoreBackend'); },
      defaultValue: 'sqlite_vec',
      options: [
        { value: 'sqlite_vec', label: i18n.t('settings:rag.opt.vectorStoreSqliteVec') },
        { value: 'chroma', label: i18n.t('settings:rag.opt.vectorStoreChroma') }
      ],
      required: true,
      get helpText() { return i18n.t('settings:rag.field.vectorStoreBackend.help'); }
    },
    {
      type: 'text',
      name: 'vector_sqlite_path',
      get label() { return i18n.t('settings:rag.field.vectorSqlitePath'); },
      get placeholder() { return i18n.t('settings:rag.field.vectorSqlitePath.placeholder'); },
      defaultValue: 'data/state/rag_vec.sqlite3',
      get helpText() { return i18n.t('settings:rag.field.vectorSqlitePath.help'); },
      condition: (formData) => formData.vector_store_backend === 'sqlite_vec',
    },
    {
      type: 'text',
      name: 'persist_directory',
      get label() { return i18n.t('settings:rag.field.chromaPersistDirectory'); },
      get placeholder() { return i18n.t('settings:rag.field.chromaPersistDirectory.placeholder'); },
      defaultValue: 'data/chromadb',
      get helpText() { return i18n.t('settings:rag.field.chromaPersistDirectory.help'); },
      condition: (formData) => formData.vector_store_backend === 'chroma',
    },
    {
      type: 'select',
      name: 'embedding_provider',
      get label() { return i18n.t('settings:rag.field.embeddingProvider'); },
      required: true,
      defaultValue: 'api',
      get options() {
        return [
          { value: 'api', label: i18n.t('settings:rag.opt.api') },
          { value: 'local', label: i18n.t('settings:rag.opt.local') },
          { value: 'local_gguf', label: i18n.t('settings:rag.opt.localGguf') }
        ];
      },
      get helpText() { return i18n.t('settings:rag.field.embeddingProvider.help'); }
    },
    {
      type: 'text',
      name: 'embedding_api_model',
      get label() { return i18n.t('settings:rag.field.embeddingApiModel'); },
      get placeholder() { return i18n.t('settings:rag.field.embeddingApiModel.placeholder'); },
      defaultValue: 'jina-embeddings-v3',
      get helpText() { return i18n.t('settings:rag.field.embeddingApiModel.help'); },
      condition: (formData) => formData.embedding_provider === 'api',
    },
    {
      type: 'text',
      name: 'embedding_api_base_url',
      get label() { return i18n.t('settings:rag.field.embeddingApiBaseUrl'); },
      get placeholder() { return i18n.t('settings:rag.field.embeddingApiBaseUrl.placeholder'); },
      defaultValue: '',
      get helpText() { return i18n.t('settings:rag.field.embeddingApiBaseUrl.help'); },
      condition: (formData) => formData.embedding_provider === 'api',
    },
    {
      type: 'password',
      name: 'embedding_api_key',
      get label() { return i18n.t('settings:rag.field.embeddingApiKey'); },
      get placeholder() { return i18n.t('settings:rag.field.embeddingApiKey.placeholder'); },
      defaultValue: '',
      get helpText() { return i18n.t('settings:rag.field.embeddingApiKey.help'); },
      condition: (formData) => formData.embedding_provider === 'api',
    },
    {
      type: 'text',
      name: 'embedding_local_model',
      get label() { return i18n.t('settings:rag.field.embeddingLocalModel'); },
      get placeholder() { return i18n.t('settings:rag.field.embeddingLocalModel.placeholder'); },
      defaultValue: 'all-MiniLM-L6-v2',
      get helpText() { return i18n.t('settings:rag.field.embeddingLocalModel.help'); },
      condition: (formData) => formData.embedding_provider === 'local',
    },
    {
      type: 'select',
      name: 'embedding_local_device',
      get label() { return i18n.t('settings:rag.field.embeddingLocalDevice'); },
      defaultValue: 'cpu',
      options: [
        { value: 'cpu', label: 'CPU' },
        { value: 'cuda', label: 'CUDA (GPU)' }
      ],
      get helpText() { return i18n.t('settings:rag.field.embeddingLocalDevice.help'); },
      condition: (formData) => formData.embedding_provider === 'local',
    },
    {
      type: 'text',
      name: 'embedding_local_gguf_model_path',
      get label() { return i18n.t('settings:rag.field.embeddingLocalGgufModelPath'); },
      get placeholder() { return i18n.t('settings:rag.field.embeddingLocalGgufModelPath.placeholder'); },
      defaultValue: 'models/embeddings/qwen3-embedding-0.6b.gguf',
      get helpText() { return i18n.t('settings:rag.field.embeddingLocalGgufModelPath.help'); },
      condition: (formData) => formData.embedding_provider === 'local_gguf',
    },
    {
      type: 'number',
      name: 'embedding_local_gguf_n_ctx',
      get label() { return i18n.t('settings:rag.field.embeddingLocalGgufCtx'); },
      min: 256,
      max: 65536,
      defaultValue: 2048,
      get helpText() { return i18n.t('settings:rag.field.embeddingLocalGgufCtx.help'); },
      condition: (formData) => formData.embedding_provider === 'local_gguf',
    },
    {
      type: 'number',
      name: 'embedding_local_gguf_n_threads',
      get label() { return i18n.t('settings:rag.field.embeddingLocalGgufThreads'); },
      min: 0,
      max: 256,
      defaultValue: 0,
      get helpText() { return i18n.t('settings:rag.field.embeddingLocalGgufThreads.help'); },
      condition: (formData) => formData.embedding_provider === 'local_gguf',
    },
    {
      type: 'number',
      name: 'embedding_local_gguf_n_gpu_layers',
      get label() { return i18n.t('settings:rag.field.embeddingLocalGgufGpuLayers'); },
      min: 0,
      max: 1024,
      defaultValue: 0,
      get helpText() { return i18n.t('settings:rag.field.embeddingLocalGgufGpuLayers.help'); },
      condition: (formData) => formData.embedding_provider === 'local_gguf',
    },
    {
      type: 'checkbox',
      name: 'embedding_local_gguf_normalize',
      get label() { return i18n.t('settings:rag.field.embeddingLocalGgufNormalize'); },
      defaultValue: true,
      get helpText() { return i18n.t('settings:rag.field.embeddingLocalGgufNormalize.help'); },
      condition: (formData) => formData.embedding_provider === 'local_gguf',
    },
    {
      type: 'number',
      name: 'embedding_batch_size',
      get label() { return i18n.t('settings:rag.field.embeddingBatchSize'); },
      min: 1,
      max: 1000,
      defaultValue: 64,
      get helpText() { return i18n.t('settings:rag.field.embeddingBatchSize.help'); }
    },
    {
      type: 'number',
      name: 'embedding_batch_delay_seconds',
      get label() { return i18n.t('settings:rag.field.embeddingBatchDelaySeconds'); },
      min: 0,
      max: 60,
      step: 0.1,
      defaultValue: 0.5,
      get helpText() { return i18n.t('settings:rag.field.embeddingBatchDelaySeconds.help'); }
    },
    {
      type: 'number',
      name: 'embedding_batch_max_retries',
      get label() { return i18n.t('settings:rag.field.embeddingBatchMaxRetries'); },
      min: 0,
      max: 20,
      defaultValue: 3,
      get helpText() { return i18n.t('settings:rag.field.embeddingBatchMaxRetries.help'); }
    },
    {
      type: 'number',
      name: 'chunk_size',
      get label() { return i18n.t('settings:rag.field.chunkSize'); },
      min: 100,
      max: 10000,
      defaultValue: 1000,
      required: true,
      get helpText() { return i18n.t('settings:rag.field.chunkSize.help'); }
    },
    {
      type: 'number',
      name: 'chunk_overlap',
      get label() { return i18n.t('settings:rag.field.chunkOverlap'); },
      min: 0,
      max: 5000,
      defaultValue: 200,
      required: true,
      get helpText() { return i18n.t('settings:rag.field.chunkOverlap.help'); }
    },
    {
      type: 'checkbox',
      name: 'query_transform_enabled',
      label: 'Enable Query Transform',
      defaultValue: false,
      helpText: 'Rewrite user query before retrieval to improve recall on complex questions.'
    },
    {
      type: 'select',
      name: 'query_transform_mode',
      label: 'Query Transform Mode',
      defaultValue: 'rewrite',
      options: [
        { value: 'none', label: 'None' },
        { value: 'rewrite', label: 'Rewrite' }
      ],
      required: true,
      condition: (formData) => formData.query_transform_enabled === true,
      helpText: 'Current MVP mode supports single-query rewrite.'
    },
    {
      type: 'text',
      name: 'query_transform_model_id',
      label: 'Query Transform Model',
      placeholder: 'auto or provider:model',
      defaultValue: 'auto',
      condition: (formData) => formData.query_transform_enabled === true,
      helpText: 'Use auto to follow the current chat model. Or set a fixed model id.'
    },
    {
      type: 'number',
      name: 'query_transform_timeout_seconds',
      label: 'Query Transform Timeout (s)',
      min: 1,
      max: 30,
      defaultValue: 4,
      required: true,
      condition: (formData) => formData.query_transform_enabled === true,
      helpText: 'If timeout or error occurs, system falls back to the original query.'
    },
    {
      type: 'checkbox',
      name: 'query_transform_guard_enabled',
      label: 'Enable Rewrite Guard',
      defaultValue: true,
      condition: (formData) => formData.query_transform_enabled === true,
      helpText: 'Block unsafe rewrites that add too many unseen terms or drop key constraints.'
    },
    {
      type: 'number',
      name: 'query_transform_guard_max_new_terms',
      label: 'Rewrite Guard Max New Terms',
      min: 0,
      max: 20,
      defaultValue: 2,
      required: true,
      condition: (formData) =>
        formData.query_transform_enabled === true && formData.query_transform_guard_enabled === true,
      helpText: 'If rewritten query introduces more than this number of new key terms, fallback to original.'
    },
    {
      type: 'checkbox',
      name: 'query_transform_crag_enabled',
      label: 'Enable CRAG Quality Gate',
      defaultValue: true,
      condition: (formData) => formData.query_transform_enabled === true,
      helpText: 'Evaluate rewritten-query retrieval quality and fallback to original query when needed.'
    },
    {
      type: 'number',
      name: 'query_transform_crag_lower_threshold',
      label: 'CRAG Lower Threshold',
      min: 0,
      max: 1,
      step: 0.05,
      defaultValue: 0.35,
      required: true,
      condition: (formData) =>
        formData.query_transform_enabled === true && formData.query_transform_crag_enabled === true,
      helpText: 'Below this score, rewritten query is treated as low quality.'
    },
    {
      type: 'number',
      name: 'query_transform_crag_upper_threshold',
      label: 'CRAG Upper Threshold',
      min: 0,
      max: 1,
      step: 0.05,
      defaultValue: 0.75,
      required: true,
      condition: (formData) =>
        formData.query_transform_enabled === true && formData.query_transform_crag_enabled === true,
      helpText: 'Between lower and upper thresholds, rewritten query is ambiguous and compared with original.'
    },
    {
      type: 'select',
      name: 'retrieval_mode',
      get label() { return i18n.t('settings:rag.field.retrievalMode'); },
      defaultValue: 'hybrid',
      get options() {
        return [
          { value: 'vector', label: i18n.t('settings:rag.opt.retrievalVector') },
          { value: 'bm25', label: i18n.t('settings:rag.opt.retrievalBm25') },
          { value: 'hybrid', label: i18n.t('settings:rag.opt.retrievalHybrid') }
        ];
      },
      required: true,
      get helpText() { return i18n.t('settings:rag.field.retrievalMode.help'); }
    },
    {
      type: 'number',
      name: 'top_k',
      get label() { return i18n.t('settings:rag.field.topK'); },
      min: 1,
      max: 50,
      defaultValue: 5,
      required: true,
      get helpText() { return i18n.t('settings:rag.field.topK.help'); }
    },
    {
      type: 'number',
      name: 'recall_k',
      get label() { return i18n.t('settings:rag.field.recallK'); },
      min: 1,
      max: 200,
      defaultValue: 20,
      required: true,
      get helpText() { return i18n.t('settings:rag.field.recallK.help'); }
    },
    {
      type: 'number',
      name: 'vector_recall_k',
      get label() { return i18n.t('settings:rag.field.vectorRecallK'); },
      min: 1,
      max: 500,
      defaultValue: 5,
      required: true,
      condition: (formData) => formData.retrieval_mode !== 'bm25',
      get helpText() { return i18n.t('settings:rag.field.vectorRecallK.help'); }
    },
    {
      type: 'number',
      name: 'bm25_recall_k',
      get label() { return i18n.t('settings:rag.field.bm25RecallK'); },
      min: 1,
      max: 500,
      defaultValue: 20,
      required: true,
      condition: (formData) => formData.retrieval_mode !== 'vector',
      get helpText() { return i18n.t('settings:rag.field.bm25RecallK.help'); }
    },
    {
      type: 'number',
      name: 'bm25_min_term_coverage',
      get label() { return i18n.t('settings:rag.field.bm25MinTermCoverage'); },
      min: 0,
      max: 1,
      step: 0.05,
      defaultValue: 0.35,
      required: true,
      condition: (formData) => formData.retrieval_mode !== 'vector',
      get helpText() { return i18n.t('settings:rag.field.bm25MinTermCoverage.help'); }
    },
    {
      type: 'number',
      name: 'fusion_top_k',
      get label() { return i18n.t('settings:rag.field.fusionTopK'); },
      min: 1,
      max: 500,
      defaultValue: 30,
      required: true,
      condition: (formData) => formData.retrieval_mode === 'hybrid',
      get helpText() { return i18n.t('settings:rag.field.fusionTopK.help'); }
    },
    {
      type: 'select',
      name: 'fusion_strategy',
      get label() { return i18n.t('settings:rag.field.fusionStrategy'); },
      defaultValue: 'rrf',
      options: [
        { value: 'rrf', label: i18n.t('settings:rag.opt.fusionRrf') }
      ],
      required: true,
      condition: (formData) => formData.retrieval_mode === 'hybrid',
      get helpText() { return i18n.t('settings:rag.field.fusionStrategy.help'); }
    },
    {
      type: 'number',
      name: 'rrf_k',
      get label() { return i18n.t('settings:rag.field.rrfK'); },
      min: 1,
      max: 500,
      defaultValue: 40,
      required: true,
      condition: (formData) => formData.retrieval_mode === 'hybrid',
      get helpText() { return i18n.t('settings:rag.field.rrfK.help'); }
    },
    {
      type: 'number',
      name: 'vector_weight',
      get label() { return i18n.t('settings:rag.field.vectorWeight'); },
      min: 0,
      max: 10,
      step: 0.05,
      defaultValue: 0.05,
      required: true,
      condition: (formData) => formData.retrieval_mode === 'hybrid',
      get helpText() { return i18n.t('settings:rag.field.vectorWeight.help'); }
    },
    {
      type: 'number',
      name: 'bm25_weight',
      get label() { return i18n.t('settings:rag.field.bm25Weight'); },
      min: 0,
      max: 10,
      step: 0.1,
      defaultValue: 1,
      required: true,
      condition: (formData) => formData.retrieval_mode === 'hybrid',
      get helpText() { return i18n.t('settings:rag.field.bm25Weight.help'); }
    },
    {
      type: 'number',
      name: 'max_per_doc',
      get label() { return i18n.t('settings:rag.field.maxPerDoc'); },
      min: 0,
      max: 20,
      defaultValue: 2,
      required: true,
      get helpText() { return i18n.t('settings:rag.field.maxPerDoc.help'); }
    },
    {
      type: 'number',
      name: 'score_threshold',
      get label() { return i18n.t('settings:rag.field.scoreThreshold'); },
      min: 0,
      max: 1,
      step: 0.05,
      defaultValue: 0.65,
      required: true,
      get helpText() { return i18n.t('settings:rag.field.scoreThreshold.help'); }
    },
    {
      type: 'select',
      name: 'reorder_strategy',
      get label() { return i18n.t('settings:rag.field.reorderStrategy'); },
      defaultValue: 'long_context',
      options: [
        { value: 'long_context', label: i18n.t('settings:rag.opt.longContext') },
        { value: 'none', label: i18n.t('settings:rag.opt.none') }
      ],
      required: true,
      get helpText() { return i18n.t('settings:rag.field.reorderStrategy.help'); }
    },
    {
      type: 'number',
      name: 'context_neighbor_window',
      label: '邻居扩展窗口',
      min: 0,
      max: 10,
      defaultValue: 0,
      required: true,
      helpText: '按 chunk_index 扩展相邻片段。0 表示关闭，2 表示命中后可补前后各 2 段。'
    },
    {
      type: 'number',
      name: 'context_neighbor_max_total',
      label: '邻居扩展上限',
      min: 0,
      max: 200,
      defaultValue: 0,
      required: true,
      helpText: '邻居扩展后的总片段上限（含原始命中）。0 表示按窗口自动估算。'
    },
    {
      type: 'number',
      name: 'context_neighbor_dedup_coverage',
      label: '邻居去重覆盖阈值',
      min: 0.5,
      max: 1,
      step: 0.01,
      defaultValue: 0.9,
      required: true,
      helpText: '候选邻居内容被已选内容覆盖到该比例时会被丢弃，越小去重越激进。'
    },
    {
      type: 'checkbox',
      name: 'rerank_enabled',
      get label() { return i18n.t('settings:rag.field.rerankEnabled'); },
      defaultValue: false,
      get helpText() { return i18n.t('settings:rag.field.rerankEnabled.help'); }
    },
    {
      type: 'text',
      name: 'rerank_api_model',
      get label() { return i18n.t('settings:rag.field.rerankApiModel'); },
      get placeholder() { return i18n.t('settings:rag.field.rerankApiModel.placeholder'); },
      defaultValue: 'jina-reranker-v2-base-multilingual',
      get helpText() { return i18n.t('settings:rag.field.rerankApiModel.help'); },
      condition: (formData) => formData.rerank_enabled === true,
    },
    {
      type: 'text',
      name: 'rerank_api_base_url',
      get label() { return i18n.t('settings:rag.field.rerankApiBaseUrl'); },
      get placeholder() { return i18n.t('settings:rag.field.rerankApiBaseUrl.placeholder'); },
      defaultValue: 'https://api.jina.ai/v1/rerank',
      get helpText() { return i18n.t('settings:rag.field.rerankApiBaseUrl.help'); },
      condition: (formData) => formData.rerank_enabled === true,
    },
    {
      type: 'password',
      name: 'rerank_api_key',
      get label() { return i18n.t('settings:rag.field.rerankApiKey'); },
      get placeholder() { return i18n.t('settings:rag.field.rerankApiKey.placeholder'); },
      defaultValue: '',
      get helpText() { return i18n.t('settings:rag.field.rerankApiKey.help'); },
      condition: (formData) => formData.rerank_enabled === true,
    },
    {
      type: 'number',
      name: 'rerank_timeout_seconds',
      get label() { return i18n.t('settings:rag.field.rerankTimeoutSeconds'); },
      min: 1,
      max: 120,
      defaultValue: 20,
      required: true,
      get helpText() { return i18n.t('settings:rag.field.rerankTimeoutSeconds.help'); },
      condition: (formData) => formData.rerank_enabled === true,
    },
    {
      type: 'number',
      name: 'rerank_weight',
      get label() { return i18n.t('settings:rag.field.rerankWeight'); },
      min: 0,
      max: 1,
      step: 0.05,
      defaultValue: 0.7,
      required: true,
      get helpText() { return i18n.t('settings:rag.field.rerankWeight.help'); },
      condition: (formData) => formData.rerank_enabled === true,
    }
  ],

  transformSave: (data: any) => {
    const rest = { ...data };
    delete rest.retrieval_preset;
    return rest;
  }
};

File: src/api/routers/rag_config.py
"""
RAG Config API Router

Provides endpoints for configuring RAG (Retrieval-Augmented Generation) settings.
"""

import logging
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from src.api.routers.service_protocols import FlatConfigServiceLike
from src.infrastructure.config.rag_config_service import RagConfigService

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/rag", tags=["rag"])


class RagConfigResponse(BaseModel):
    """Response model for RAG configuration"""

    embedding_provider: str
    embedding_api_model: str
    embedding_api_base_url: str = ""
    embedding_api_key: str = ""
    embedding_local_model: str
    embedding_local_device: str
    embedding_local_gguf_model_path: str
    embedding_local_gguf_n_ctx: int
    embedding_local_gguf_n_threads: int
    embedding_local_gguf_n_gpu_layers: int
    embedding_local_gguf_normalize: bool
    embedding_batch_size: int
    embedding_batch_delay_seconds: float
    embedding_batch_max_retries: int
    chunk_size: int
    chunk_overlap: int
    retrieval_mode: str
    top_k: int
    score_threshold: float
    recall_k: int
    vector_recall_k: int
    bm25_recall_k: int
    bm25_min_term_coverage: float
    fusion_top_k: int
    fusion_strategy: str
    rrf_k: int
    vector_weight: float
    bm25_weight: float
    max_per_doc: int
    reorder_strategy: str
    context_neighbor_window: int
    context_neighbor_max_total: int
    context_neighbor_dedup_coverage: float
    retrieval_query_planner_enabled: bool
    retrieval_query_planner_model_id: str
    retrieval_query_planner_max_queries: int
    retrieval_query_planner_timeout_seconds: int
    structured_source_context_enabled: bool
    query_transform_enabled: bool
    query_transform_mode: str
    query_transform_model_id: str
    query_transform_timeout_seconds: int
    query_transform_guard_enabled: bool
    query_transform_guard_max_new_terms: int
    query_transform_crag_enabled: bool
    query_transform_crag_lower_threshold: float
    query_transform_crag_upper_threshold: float
    rerank_enabled: bool
    rerank_api_model: str
    rerank_api_base_url: str
    rerank_api_key: str = ""
    rerank_timeout_seconds: int
    rerank_weight: float
    vector_store_backend: str
    vector_sqlite_path: str
    persist_directory: str
    bm25_sqlite_path: str


class RagConfigUpdate(BaseModel):
    """Request model for updating RAG configuration"""

    embedding_provider: str | None = None
    embedding_api_model: str | None = None
    embedding_api_base_url: str | None = None
    embedding_api_key: str | None = None
    embedding_local_model: str | None = None
    embedding_local_device: str | None = None
    embedding_local_gguf_model_path: str | None = None
    embedding_local_gguf_n_ctx: int | None = Field(default=None, ge=256, le=65536)
    embedding_local_gguf_n_threads: int | None = Field(default=None, ge=0, le=256)
    embedding_local_gguf_n_gpu_layers: int | None = Field(default=None, ge=0, le=1024)
    embedding_local_gguf_normalize: bool | None = None
    embedding_batch_size: int | None = Field(default=None, ge=1, le=1000)
    embedding_batch_delay_seconds: float | None = Field(default=None, ge=0.0, le=60.0)
    embedding_batch_max_retries: int | None = Field(default=None, ge=0, le=20)
    chunk_size: int | None = Field(default=None, ge=100, le=10000)
    chunk_overlap: int | None = Field(default=None, ge=0, le=5000)
    retrieval_mode: Literal["vector", "bm25", "hybrid"] | None = None
    top_k: int | None = Field(default=None, ge=1, le=50)
    score_threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    recall_k: int | None = Field(default=None, ge=1, le=200)
    vector_recall_k: int | None = Field(default=None, ge=1, le=500)
    bm25_recall_k: int | None = Field(default=None, ge=1, le=500)
    bm25_min_term_coverage: float | None = Field(default=None, ge=0.0, le=1.0)
    fusion_top_k: int | None = Field(default=None, ge=1, le=500)
    fusion_strategy: Literal["rrf"] | None = None
    rrf_k: int | None = Field(default=None, ge=1, le=500)
    vector_weight: float | None = Field(default=None, ge=0.0, le=10.0)
    bm25_weight: float | None = Field(default=None, ge=0.0, le=10.0)
    max_per_doc: int | None = Field(default=None, ge=0, le=20)
    reorder_strategy: Literal["none", "long_context"] | None = None
    context_neighbor_window: int | None = Field(default=None, ge=0, le=10)
    context_neighbor_max_total: int | None = Field(default=None, ge=0, le=200)
    context_neighbor_dedup_coverage: float | None = Field(default=None, ge=0.5, le=1.0)
    retrieval_query_planner_enabled: bool | None = None
    retrieval_query_planner_model_id: str | None = None
    retrieval_query_planner_max_queries: int | None = Field(default=None, ge=1, le=8)
    retrieval_query_planner_timeout_seconds: int | None = Field(default=None, ge=1, le=30)
    structured_source_context_enabled: bool | None = None
    query_transform_enabled: bool | None = None
    query_transform_mode: Literal["none", "rewrite"] | None = None
    query_transform_model_id: str | None = None
    query_transform_timeout_seconds: int | None = Field(default=None, ge=1, le=30)
    query_transform_guard_enabled: bool | None = None
    query_transform_guard_max_new_terms: int | None = Field(default=None, ge=0, le=20)
    query_transform_crag_enabled: bool | None = None
    query_transform_crag_lower_threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    query_transform_crag_upper_threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    rerank_enabled: bool | None = None
    rerank_api_model: str | None = None
    rerank_api_base_url: str | None = None
    rerank_api_key: str | None = None
    rerank_timeout_seconds: int | None = Field(default=None, ge=1, le=120)
    rerank_weight: float | None = Field(default=None, ge=0.0, le=1.0)
    vector_store_backend: Literal["sqlite_vec", "chroma"] | None = None
    vector_sqlite_path: str | None = None
    persist_directory: str | None = None
    bm25_sqlite_path: str | None = None


def get_rag_config_service() -> RagConfigService:
    """Dependency injection for RagConfigService."""
    return RagConfigService()


@router.get("/config", response_model=RagConfigResponse)
async def get_config(service: FlatConfigServiceLike = Depends(get_rag_config_service)):
    """Get current RAG configuration"""
    try:
        flat = service.get_flat_config()
        return RagConfigResponse(**flat)
    except Exception as e:
        logger.error(f"Failed to get RAG config: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/config")
async def update_config(
    updates: RagConfigUpdate, service: FlatConfigServiceLike = Depends(get_rag_config_service)
):
    """Update RAG configuration"""
    try:
        update_dict = updates.model_dump(exclude_none=True)
        if not update_dict:
            raise HTTPException(status_code=400, detail="No updates provided")

        # Validate embedding provider
        if "embedding_provider" in update_dict:
            allowed = {"api", "local", "local_gguf"}
            if update_dict["embedding_provider"] not in allowed:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported embedding provider: {update_dict['embedding_provider']}",
                )
        if "reorder_strategy" in update_dict:
            allowed_strategies = {"none", "long_context"}
            if update_dict["reorder_strategy"] not in allowed_strategies:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported reorder strategy: {update_dict['reorder_strategy']}",
                )
        if "query_transform_mode" in update_dict:
            allowed_modes = {"none", "rewrite"}
            if update_dict["query_transform_mode"] not in allowed_modes:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported query transform mode: {update_dict['query_transform_mode']}",
                )
        if (
            "query_transform_crag_lower_threshold" in update_dict
            or "query_transform_crag_upper_threshold" in update_dict
        ):
            current = service.get_flat_config()
            lower_value = float(
                update_dict.get(
                    "query_transform_crag_lower_threshold",
                    current["query_transform_crag_lower_threshold"],
                )
            )
            upper_value = float(
                update_dict.get(
                    "query_transform_crag_upper_threshold",
                    current["query_transform_crag_upper_threshold"],
                )
            )
            if lower_value >= upper_value:
                raise HTTPException(
                    status_code=400,
                    detail="query_transform_crag_lower_threshold must be smaller than query_transform_crag_upper_threshold",
                )
        if "fusion_strategy" in update_dict:
            allowed_fusion = {"rrf"}
            if update_dict["fusion_strategy"] not in allowed_fusion:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported fusion strategy: {update_dict['fusion_strategy']}",
                )
        if "vector_store_backend" in update_dict:
            allowed_backends = {"sqlite_vec", "chroma"}
            if update_dict["vector_store_backend"] not in allowed_backends:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported vector backend: {update_dict['vector_store_backend']}",
                )

        service.save_flat_config(update_dict)
        return {"message": "RAG configuration updated successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to update RAG config: {e}")
        raise HTTPException(status_code=500, detail=str(e))
